### ISSUE TITLE
Add support for exporting into NWBv2 as well

### ISSUE THREAD

**Issue by t-b** [AllenInstitute/MIES#22]
#### Cross links

https://github.com/AllenInstitute/MIES/issues/22
https://github.com/AllenInstitute/MIES/pull/179 (TODO list)
https://github.com/AllenInstitute/IPNWB/issues/5
https://github.com/AllenInstitute/IPNWB/pull/16

Support matrix (read/write):

Igor Pro Version \ NWB Version| v1 | v2
------------ | -------------| -------------
8 | x | x

(IP7 is not supported anymore)

Tests:
- [x] Copy UTF_NwbExportV1.ipf, adapt for V2 and test both exports in all current cases
- [x] Add validation tests using pynwb for V2 files as well
- [x] Update UTF_NwbExportV2.ipf with changes to UTF_NwbExportV1.ipf since last sync

**Comment by t-b** [comment]
Work here should start on top of the branch `bugfix/fixup-buggy-headstage-states-on-export`.

**Comment by t-b** [comment]
Support matrix:

Igor Pro Version \ NWB Version| v1 | v2
------------ | -------------| -------------
7  | x | o
8 | x | x

**Comment by t-b** [comment]
@ukos-git pynwb validation: https://pynwb.readthedocs.io/en/stable/validation.html

**Comment by ukos-git** [comment]
validation [fails in pynwb here](https://github.com/NeurodataWithoutBorders/pynwb/blob/a081b39803f0ffb535b2198a2fdd81007ffa442e/src/pynwb/__init__.py#L183) due to an error at HDF5IO (in hdftools.py) at [h5py level](https://github.com/h5py/h5py/blob/ab949c95966595f59b30c6f1db15c466a4bbae77/h5py/_hl/attrs.py#L81) when reading the `IGORWaveNote` attribute.

<details>
  <summary>Stack trace of the above error:</summary>
<pre>
python -m pynwb.validate ~/Documents/Playground/mies-playground/nwbv2/functionanalysis-compressed.nwb  
Validating /home/matthias/Documents/Playground/mies-playground/nwbv2/functionanalysis-compressed.nwb against core namespace
Traceback (most recent call last):
  File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/matthias/Documents/Projects/pynwb/src/pynwb/validate.py", line 71, in <module>
    main()
  File "/home/matthias/Documents/Projects/pynwb/src/pynwb/validate.py", line 65, in main
    ret = ret or _validate_helper(io=io)
  File "/home/matthias/Documents/Projects/pynwb/src/pynwb/validate.py", line 23, in _validate_helper
    errors = validate(**kwargs)
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/utils.py", line 396, in func_call
    return func(**parsed['args'])
  File "/home/matthias/Documents/Projects/pynwb/src/pynwb/__init__.py", line 183, in validate
    builder = io.read_builder()
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/utils.py", line 381, in func_call
    return func(self, **parsed['args'])
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/backends/hdf5/h5tools.py", line 227, in read_builder
    f_builder = self.__read_group(self.__file, ROOT_NAME, ignore=ignore)
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/backends/hdf5/h5tools.py", line 302, in __read_group
    builder = read_method(sub_h5obj)
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/backends/hdf5/h5tools.py", line 302, in __read_group
    builder = read_method(sub_h5obj)
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/backends/hdf5/h5tools.py", line 302, in __read_group
    builder = read_method(sub_h5obj)
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/backends/hdf5/h5tools.py", line 316, in __read_dataset
    "attributes": self.__read_attrs(h5obj),
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/hdmf/backends/hdf5/h5tools.py", line 369, in __read_attrs
    for k, v in h5obj.attrs.items():
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/h5py/_hl/base.py", line 353, in __iter__
    yield (key, self._mapping.get(key))
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/_collections_abc.py", line 660, in get
    return self[key]
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "/home/matthias/Documents/Projects/pynwb/virtualenv/lib/python3.7/site-packages/h5py/_hl/attrs.py", line 81, in __getitem__
    attr.read(arr, mtype=htype)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5a.pyx", line 355, in h5py.h5a.AttrID.read
  File "h5py/_proxy.pyx", line 36, in h5py._proxy.attr_rw
OSError: Unable to read attribute (no appropriate function for conversion path)
</pre>
</details>

I am guessing, there is something wrong with this attribute implementation on Igor Pro site but I can be mistaken. I am not sure if skipping IGOR specific wave notes with format extensions could would work. I think we need to create a mwe HDF5 file with a Igor Wave Note and check. Probably need to report to WaveMetrics if invalid HDF5 format is generated upstream.

I am skipping pynwb validation for now.

**Comment by t-b** [comment]
@ukos-git Skip it for now, can you post an nwbv2 here as well? I might take a look.

**Comment by ukos-git** [comment]
[functionanalysis-compressed.zip](https://github.com/AllenInstitute/MIES/files/3219506/functionanalysis-compressed.zip)

**Comment by t-b** [comment]
@ukos-git Thanks. As I thought this is a known h5py issue, see https://github.com/h5py/h5py/pull/988. I thought they had fixed it in a PR but I just tried with h5py 2.9.0 and has the same issue.

For the development you can skip writing these Igor attributes by forcing writeIgorAttr to 0 in

```
$ git grep -n "writeIgorAttr = "
IPNWB_HDF5Helpers.ipf:40:       writeIgorAttr = ParamIsDefault(writeIgorAttr) ? 0 : !!writeIgorAttr
IPNWB_HDF5Helpers.ipf:77:       writeIgorAttr = ParamIsDefault(writeIgorAttr) ? 0 : !!writeIgorAttr
```

and then validation should be possible.

**Comment by t-b** [comment]
@ukos-git DynamicTable/Ragged Array explanations in the paper: https://www.biorxiv.org/content/10.1101/523035v1
